src/main/java/pl/damianszczepanik/jenkins/buildhistorymanager/model/actions/DeleteArtifactsWithPatternAction.java

```diff
@@ -0,0 +1,175 @@
+        public Void invoke(File vFile, VirtualChannel channel) throws IOException {          
+
+            Set<File> directories = new HashSet<>();
+            deleteFileOrLogError(vFile, directories);
+            deleteEmptyDirectoriesAndParents(directories);
+            return null;
+        }
+
+        public void deleteEmptyDirectoriesAndParents(Set<File> directories) throws IOException {
+            for (File dir : directories){
+                if (shouldDeleteDirectory(dir)) {
+                    Util.deleteFile(dir);
+                    deleteParentDirectories(dir.getParentFile());
+                }
+            }
+        }
+
+        public boolean shouldDeleteDirectory(File dir) throws IOException {
+            if (isInvalidDirectory(dir)) {
+                return false;
+            }
+
+            File parent = dir.getParentFile();
+            return hasValidParent(parent);
+        }
+
+        public boolean hasValidParent(File parent) {
+            return parent != null && parent.getPath().equals(this.archiveRootPath);
+        }
+
+        public boolean isInvalidDirectory(File dir) throws IOException {
+            return dir == null || !dir.isDirectory() || !isDirEmpty(dir.toPath());
+        }
+
+        public void deleteParentDirectories(File directory) {
+            File parent = directory;
+
+            while (shouldDelete(parent)) {
+                deleteDirectory(parent);
+                parent = parent.getParentFile();
+            }
+        }
+
+        public boolean shouldDelete(File directory) {
+            return directory != null && !directory.getPath().equals(this.archiveRootPath);
+        }
+
+        public void deleteDirectory(File directory) {
+            boolean deleteSuccess = directory.delete();
+            if(!deleteSuccess) {
+                throw new RuntimeException("Deletion of directory failed: " + directory.getAbsolutePath());
+            }
+        }
+
+        public void deleteFileOrLogError(File vFile, Set<File> directories) throws IOException {
+            if (vFile.isFile()) {
+                LOGGER.log(Level.FINE, "Deleting " + vFile.getName());
+                directories.add(vFile.getParentFile());
+                Util.deleteFile(vFile);
+                return; // Return early when the condition is met
+            }
+            LOGGER.log(Level.FINE, vFile + " is neither a directory nor a regular file");
+        }
+
+        // checkRoles method is used for access control and security purposes in Jenkins.
+        // It is responsible for checking if the current user has the required permissions to execute the code defined in the FileCallable object.
+        // By implementing the checkRoles method and checking for the required permission, the code is more secure and protected against unauthorized access.
+        @Override
+        public void checkRoles(RoleChecker checker) throws SecurityException {
+            // Nothing to do here  
```

Comment:
and Jenkins expects that this method is implemented anyway ?

Reply:
vscode auto-generated this method to make the plugin work.

Reply:
why, for what? other implementation do not provide this method so I don't think this is right way

Reply:
I got the following error when I removed `checkRoles` method. 
![image](https://github.com/jenkinsci/build-history-manager-plugin/assets/90399606/c2b36fc7-dc9c-4aa6-9747-29e151623693)
It looks like we need to add this method in order to use `FileCallable`

It is documented here:
https://javadoc.jenkins.io/hudson/FilePath.FileCallable.html